migrate structure tests over to structure test framework

nkubala opened the pull request:
also, add container build artifacts. The current Jenkins build job will need to be changed to use the new jenkins_build.sh script.

dependent on GoogleCloudPlatform/runtimes-common#49.

nkubala commented:
@dlorenc @broady

broady commented:
```
$ ./ext_run.sh -i gcr.io/google_appengine/golang -v --config test_config.json
/home/travis/build.sh: line 47: ./ext_run.sh: Permission denied
```
cc @zombiezen

nkubala commented:
ah, this is the first time i've actually tried running these in travis. just added a `chmod +x` to the travis yaml, let's see if this run passes

zombiezen commented on .travis.yml:
```diff
@@ -8,4 +8,6 @@ branches:
 
 script:
-- ./test.bash
+- curl https://raw.githubusercontent.com/GoogleCloudPlatform/runtimes-common/master/structure_tests/ext_run.sh > ext_run.sh
```
Can you pin this to a specific commit?  I like reproducible builds/tests, although I realize that every other part of this build can slide versions around.

nkubala replied:
I suppose you could pin this to a specific commit hash, but in general we don't really want people using anything except master/latest. We'll be releasing new versions of the image continuously and be able to rollback should anything break. But if you really wanted to nothing is stopping you from replacing `master` with a commit hash you know works.

zombiezen commented on base/cloudbuild.yaml.in:
```diff
@@ -0,0 +1,8 @@
+steps:
```
Similar to above, can we pin these images to a particular tag?

nkubala replied:
Same idea as above, but in this case you're pulling an image from GCR rather than code directly from github, so to keep things consistent it's probably safer to stay on master (since we won't actually be tagging any other images besides `latest`, you'd need to find the specific digest of the image you were using). If we break something then it's easier for us just to rollback rather than have you guys try and work around it, but if you absolutely need to then the option is there.

broady commented on build.sh:
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash
+
+usage() { echo "Usage: ./build.sh [target_image_path]"; exit 1; }
```
What's an example of a target_image_path?

Where is build.sh called from?

nkubala replied:
an example run of this script would be
 ```./build.sh gcr.io/google_appengine/golang-docker```
currently build.sh exists mostly for convenience (i.e. running locally) just as a wrapper around a cloud build call.

broady commented on base/cloudbuild.yaml.in:
```diff
@@ -0,0 +1,8 @@
+steps:
+  - name: gcr.io/cloud-builders/docker
+    args: ['build', '--tag=${IMAGE}', './base']
```
Shouldn't `IMAGE` be named `TAG`?

nkubala replied:
possibly, but I think that's more an issue with `docker build --tag` (since the value provided is technically the entire image name plus the tag, not just the tag itself). the `IMAGE` value is used again after this when it's provided to the `--image` flag of the structure tests, which makes more sense.

broady commented on base/cloudbuild.yaml.in:
```diff
@@ -0,0 +1,8 @@
+    args: ['build', '--tag=${IMAGE}', './base']
+    id: BUILD
+  - name: gcr.io/gcp-runtimes/structure_test
+    args: ['--image', '${IMAGE}', '-v', '--config', '/workspace/test/test_config.json']
```
Can't you pass a relative path here?

`./test/test_config.json`

nkubala replied:
since this is a cloud build step, this is run inside of an intermediate container that's constructed from all the previous steps. as per the cloud build spec, everything in the project is copied into `/workspace` inside this container, so we have to pass the full path since the structure test binary is also running inside of this same container.

nkubala replied:
ah i see what you mean. that should work

nkubala replied:
ah nevermind actually. this is an argument being passed to the structure test binary, which uses a Go library to read in the config file. so it needs to be the fully qualified path.

broady commented on .travis.yml:
```diff
@@ -8,4 +8,6 @@ branches:
-- ./test.bash
+- curl https://raw.githubusercontent.com/GoogleCloudPlatform/runtimes-common/master/structure_tests/ext_run.sh > ext_run.sh
+- chmod +x ext_run.sh
+- ./ext_run.sh -i gcr.io/google_appengine/golang -v --config test_config.json
```
`test/test_config.json`, right?

nkubala replied:
good catch.

broady commented on jenkins_build.sh:
```diff
@@ -0,0 +1,20 @@
+set -e
```
File level comment, please. What is this for?

zombiezen replied:
This is a pretty common shell script trick to add the equivalent of:

if [[ $? -ne 0 ]]; then exit 1; fi

Think of it as implicitly adding the `if err != nil` check in Go.

broady replied:
Not this line specifically, Ross. The entire file.

zombiezen replied:
@broady Oops, sorry, I misunderstood.

nkubala replied:
i'm fine with changing this however you guys see fit, or just removing entirely. i assume it's just here to halt the jenkins build in case any part of this script fails (so nothing weird happens when calling gcloud commands or something)

broady commented:
I think there's something very wrong with the test script.

```
The command "chmod +x ext_run.sh" exited with 0.
19.82s$ ./ext_run.sh -i gcr.io/google_appengine/golang -v --config test/test_config.json
Using default tag: latest
latest: Pulling from gcp-runtimes/structure_test
Status: Downloaded newer image for gcr.io/gcp-runtimes/structure_test:latest
Unable to find image 'gcr.io/google_appengine/golang:latest' locally
latest: Pulling from google_appengine/golang
Status: Downloaded newer image for gcr.io/google_appengine/golang:latest
docker: Error response from daemon: oci runtime error: exec: "./test/structure_test": stat ./test/structure_test: no such file or directory.
```

Two things:

* The image should be local. Testing the image that's on gcr.io shouldn't be what's happening.
* The script appears to have failed but didn't return a non-zero exit code.

/cc @dlorenc

nkubala commented:
the script will fail with that error until https://github.com/GoogleCloudPlatform/runtimes-common/pull/49 is merged and deployed; in the current version of the image there is no volume to mount in the image under test.

the image needs to get on the local machine somehow, so it can either be pulled from gcr.io or built by the user. building the image manually (by cloning the https://github.com/GoogleCloudPlatform/runtimes-common repository and doing a docker build) and then passing the `--no-pull` flag to the script will accomplish this.

i'll look into why the script exited with 0 after it failed.

broady commented:
@nkubala there's no set -e in run_ext

also, yes, the testing image needs to be pulled, then golang-docker image should *not* be pulled.

We should never see this in the logs:

```
Unable to find image 'gcr.io/google_appengine/golang:latest' locally
latest: Pulling from google_appengine/golang
```

nkubala commented:
alright, this should be fixed now. there were a few issues here, the biggest of which being that the entrypoint in the ext_run.sh script was failing if the container was started in any directory other than root (didn't catch this because golang is the first image i've seen to start somewhere else). fixed this and added failfast to the script here: https://github.com/GoogleCloudPlatform/runtimes-common/pull/54.

additionally, because of the way the travis yaml was set up, the tests were being run on the latest image pushed to the repository, instead of the image built off of the newly staged code. fixed the travis script to first build the image locally, then run the script on top of this image.

this PR does still depend on GoogleCloudPlatform/runtimes-common#49, but for a different reason than I had originally mentioned: this PR adds support for environment variable substitution in the tests, which is necessary for these tests to pass.

sorry for the confusion. cc @broady @dlorenc

broady commented:
This looks very close now. You can change the test to check the output of GOPATH in its entirety.

It kind of unfortunate the tests can't have comments because they're written in JSON. Any though to using YAML? (@dlorenc)

dlorenc commented:
> It kind of unfortunate the tests can't have comments because they're written in JSON. Any though to using YAML? (@dlorenc)

Yeah it already supports YAML too. Nick, could you switch this PR over to the YAML version?

nkubala commented:
> Nick, could you switch this PR over to the YAML version?
done.

> You can change the test to check the output of GOPATH in its entirety.
done.

zombiezen commented:
:+1: LGTM. Thanks so much for working on this, Nick!

broady commented:
```
2016/11/09 21:19:03 Error parsing config file: invalid character 's' looking for beginning of value
```

nkubala commented:
> ```2016/11/09 21:19:03 Error parsing config file: invalid character 's' looking for beginning```

https://github.com/GoogleCloudPlatform/runtimes-common/pull/56
